_posts/2018-05-31-regular-types.md

```diff
@@ -0,0 +1,874 @@
+
+const T a = SomeT();
+const T b = SomeT();
+if (a == b) {
+  DoSomething(a);
+  assert(a == b);
+}
+```
+
+Is it really the case that `DoSomething()` can do anything we want and leave
+this snippet intact? For an `int*`, it definitely isn't - certainly not to the
+same extent as it is for `int`. (There is perhaps some overlap between this and
+Lisa Lippincott's recent work [What is The Basic
+Interface](https://www.youtube.com/watch?v=s70b2P3A3lg).)
+
+For instance, if we dereference the pointer, we may have introduced a data race:
+`int*` is only thread-compatible if never dereferenced, but we do not have
+knowledge that nothing else is modifying the underlying `int`. Or worse, if we
+dereference and **write** to the pointer, we make it a race if any thread is
+even reading the `int`.
+
+```c++
+void DoSomething(int* const t) { std::cout << *t << std::endl; }
+```
+
+P0898r2 gets at some of this in [concepts.lib.general.equality] p3: "Expressions
+required by this specification to be equality preserving are further required to
+be stable: two evaluations of such an expression with the same input objects
+must have equal outputs absent any explicit intervening modification of those
+input objects. [ Note: This requirement allows generic code to reason about the
+current values of objects based on knowledge of the prior values as observed via
+equality preserving expressions. **It effectively forbids spontaneous changes to
+an object, changes to an object from another thread of execution, changes to an
+object as side effects of non-modifying expressions, and changes to an object as
+side effects of modifying a distinct object if those changes could be observable
+to a library function via an equality preserving expression that is required to
+be valid for that object**. — end note ]"
+
+In order to reason about a type that has dependent preconditions (like `int*`),
+to use it in standard algorithms, or to do much of anything with the type, we
+must have additional knowledge: why do we know that a given instance is being
+used in a race-free fashion?
+
+The same is implicitly true for traditional Regular types. When handed a
+newly-constructed `std::string`, you know everything you need to know in order
+to operate on that object safely, even in the face of data races. If you don't
+pass it to another thread, it isn't shared. You are data race free purely by
+virtue of having the object and knowing there are no additional (mutable)
+references to it anywhere. If we instead only hand you a `std::string&` (as is
+the case for basically all usage of standard algorithms), we don't actually have
+that knowledge. Without knowledge that it is unshared, you cannot compare it or
+copy it in a race free fashion.
+
+This requirement is unstated, and pervasive. Everything in the Stepanov-era
+standard library implicitly assumes it. In the Ranges/Concepts era, we get text
+like P0898, rightly forbidding spontaneous changes to an object. But in the
```

Comment:
Update this last sentence from the text in the CL.

Reply:
Done